substrate/frame/nomination-pools/src/migration.rs

```diff
@@ -22,56 +22,246 @@ use sp_std::{collections::btree_map::BTreeMap, vec::Vec};
 
 #[cfg(feature = "try-runtime")]
 use sp_runtime::TryRuntimeError;
+pub mod v6 {
+	use super::*;
 
-pub mod v1 {
+	/// Wrapper over `MigrateToV6` with convenience version checks.
+	pub type VersionedMigrateV6<T> = frame_support::migrations::VersionedMigration<
+		5,
+		6,
+		MigrateToV6<T>,
+		crate::pallet::Pallet<T>,
+		<T as frame_system::Config>::DbWeight,
+	>;
+
+	/// This migration would restrict reward account of pools to go below ED by doing a named
+	/// freeze on all the existing pools.
+	pub struct MigrateToV6<T>(sp_std::marker::PhantomData<T>);
+
+	impl<T: Config> MigrateToV6<T> {
+		fn freeze_ed(pool_id: PoolId) -> Result<(), ()> {
+			let reward_acc = Pallet::<T>::create_reward_account(pool_id);
+			Pallet::<T>::freeze_min_balance(&reward_acc).map_err(|e| {
+				log!(error, "Failed to freeze ED for pool {} with error: {:?}", pool_id, e);
+				()
+			})
+		}
+	}
+	impl<T: Config> OnRuntimeUpgrade for MigrateToV6<T> {
+		fn on_runtime_upgrade() -> Weight {
+			let mut success = 0u64;
+			let mut fail = 0u64;
+
+			BondedPools::<T>::iter_keys().for_each(|p| {
+				if Self::freeze_ed(p).is_ok() {
+					success.saturating_inc();
+				} else {
+					fail.saturating_inc();
+				}
+			});
+
+			log!(
+				info,
+				"Freezing ED succeeded for {} pools, and failed for {} pools",
+				success,
+				fail
+			);
```

Comment:
Should these ever fail? If not, maybe log an err here? If yes, maybe document under which scenarios they might fail?

Reply:
No it should never fail. We already log error for each failure. Adding failure count seemed harmless so I left it there. But great suggestion, I could make this an error log as well if there are failures.